From 4582c857e494e239d9d93a5eb46493e9da197115 Mon Sep 17 00:00:00 2001
Message-Id: <4582c857e494e239d9d93a5eb46493e9da197115.1421709247.git.chang-joon.lee@intel.com>
In-Reply-To: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
References: <3650eb44734ad4fbe098313349da8185595a20d8.1421709247.git.chang-joon.lee@intel.com>
From: Pallavi <pallavi.g@intel.com>
Date: Wed, 1 Oct 2014 14:57:41 +0530
Subject: [PATCH 50/75] MUST_REBASE [VPG]: adf/intel/vlv: Add tiling support

Added tiling support to the intel buffera and remove hardcoded tiling mode as
TILED rather than using computed variable

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

For: GMINL-3547
Change-Id: I54df63dbe8f4f979b86892192c30d3cf5366e788
Signed-off-by: Pallavi <pallavi.g@intel.com>
Signed-off-by: Shobhit Kumar <shobhit.kumar@intel.com>
---
 drivers/video/adf/intel/core/vlv/vlv_pri_plane.c   |    8 +++++---
 drivers/video/adf/intel/core/vlv/vlv_sp_plane.c    |   11 +++++------
 .../video/adf/intel/include/core/intel_dc_config.h |    1 +
 drivers/video/adf/intel/intel_adf_device.c         |    3 +++
 4 files changed, 14 insertions(+), 9 deletions(-)

diff --git a/drivers/video/adf/intel/core/vlv/vlv_pri_plane.c b/drivers/video/adf/intel/core/vlv/vlv_pri_plane.c
index 8383530..f2c04ee 100644
--- a/drivers/video/adf/intel/core/vlv/vlv_pri_plane.c
+++ b/drivers/video/adf/intel/core/vlv/vlv_pri_plane.c
@@ -214,15 +214,17 @@ static void vlv_pri_flip(struct intel_plane *plane,
 	regs->dspcntr &= ~DISPPLANE_PIXFORMAT_MASK;
 	regs->dspcntr |= format_config;
 
-	/* TODO tile support should be add as apart of the buf param*/
-	regs->dspcntr &= ~DISPPLANE_TILED;
+	if (buf->tiling_mode != I915_TILING_NONE)
+		regs->dspcntr |= DISPPLANE_TILED;
+	else
+		regs->dspcntr &= ~DISPPLANE_TILED;
 
 	regs->stride = buf->stride;
 
 	regs->linearoff = src_y * regs->stride + src_x * bpp;
 
 	dspaddr_offset = vlv_compute_page_offset(&src_x, &src_y,
-				I915_TILING_NONE, bpp, regs->stride);
+				buf->tiling_mode, bpp, regs->stride);
 
 	regs->linearoff -= dspaddr_offset;
 
diff --git a/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c b/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c
index 22b9c37..8d3ebce 100644
--- a/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c
+++ b/drivers/video/adf/intel/core/vlv/vlv_sp_plane.c
@@ -239,9 +239,6 @@ static int vlv_sp_validate(struct intel_plane *plane, struct intel_buffer *buf,
 	bool visible = false;
 	struct rectangle clip;
 
-	/* TODO need to add tiling support */
-	u32 tiling_mode = I915_TILING_NONE;
-
 	struct rectangle src = {
 		/* sample coordinates in 16.16 fixed point */
 		.x1 = config->src_x,
@@ -283,7 +280,8 @@ static int vlv_sp_validate(struct intel_plane *plane, struct intel_buffer *buf,
 	}
 
 	/* sprite planes can be linear or x-tiled surfaces */
-	if (tiling_mode != I915_TILING_NONE && tiling_mode != I915_TILING_X) {
+	if (buf->tiling_mode != I915_TILING_NONE &&
+		buf->tiling_mode != I915_TILING_X) {
 		pr_err("ADF: unsupported tiling mode %s\n", __func__);
 		return -EINVAL;
 	}
@@ -353,14 +351,15 @@ static void vlv_sp_flip(struct intel_plane *planeptr, struct intel_buffer *buf,
 	sprctl |= hw_format;
 	sprctl |= SP_GAMMA_ENABLE;
 
-	/* tile mode need to be supported */
+	if (buf->tiling_mode != I915_TILING_NONE)
+		sprctl |= SP_TILED;
 
 	sprctl |= SP_ENABLE;
 	regs->dspcntr = sprctl;
 
 	linear_offset = src_y * buf->stride + src_x * bpp;
 	sprsurf_offset = vlv_compute_page_offset(&src_x, &src_y,
-			I915_TILING_NONE, bpp, buf->stride);
+			buf->tiling_mode, bpp, buf->stride);
 	regs->linearoff = linear_offset - sprsurf_offset;
 
 	regs->stride = buf->stride;
diff --git a/drivers/video/adf/intel/include/core/intel_dc_config.h b/drivers/video/adf/intel/include/core/intel_dc_config.h
index bf7495b..e6bc4e8 100644
--- a/drivers/video/adf/intel/include/core/intel_dc_config.h
+++ b/drivers/video/adf/intel/include/core/intel_dc_config.h
@@ -107,6 +107,7 @@ struct intel_buffer {
 	u32 format;
 	unsigned long gtt_offset_in_pages;
 	u32 stride;
+	u32 tiling_mode;
 };
 
 struct intel_dc_buffer {
diff --git a/drivers/video/adf/intel/intel_adf_device.c b/drivers/video/adf/intel/intel_adf_device.c
index 0eef9e1..bfcc008 100644
--- a/drivers/video/adf/intel/intel_adf_device.c
+++ b/drivers/video/adf/intel/intel_adf_device.c
@@ -174,12 +174,14 @@ static int adf_buffer_to_intel_buffer(struct adf_buffer *adf_buf,
 	struct adf_buffer_mapping *mapping, struct intel_buffer *intel_buf)
 {
 	u32 gtt_in_pages = 0;
+	u32 tiling_mode = 0;
 #ifdef CONFIG_ADF_INTEL_VLV
 	struct dma_buf_attachment *buf_attach = mapping->attachments[0];
 	struct i915_drm_dmabuf_attachment *i915_buf_attach =
 		(struct i915_drm_dmabuf_attachment *)buf_attach->priv;
 
 	gtt_in_pages = i915_buf_attach->gtt_offset;
+	tiling_mode = i915_buf_attach->tiling_mode;
 #else
 	struct dma_buf *dma_buf = adf_buf->dma_bufs[0];
 	int err;
@@ -193,6 +195,7 @@ static int adf_buffer_to_intel_buffer(struct adf_buffer *adf_buf,
 	intel_buf->h = adf_buf->h;
 	intel_buf->gtt_offset_in_pages = gtt_in_pages;
 	intel_buf->stride = adf_buf->pitch[0];
+	intel_buf->tiling_mode = tiling_mode;
 
 	return 0;
 }
-- 
1.7.9.5

